flocker/node/agents/blockdevice.py

```diff
@@ -0,0 +1,270 @@
+# -*- test-case-name: flocker.node.agents.test.test_blockdevice -*-
+# Copyright Hybrid Logic Ltd.  See LICENSE file for details.
+
+"""
+This module implements the parts of a block-device based dataset
+convergence agent that can be re-used against many different kinds of block
+devices.
+"""
+
+from uuid import uuid4
+from subprocess import check_output
+
+from zope.interface import implementer, Interface
+
+from characteristic import attributes
+from pyrsistent import PRecord, field
+
+from twisted.python.filepath import FilePath
+
+
+class VolumeException(Exception):
+    """
+    A base class for exceptions raised by  ``IBlockDeviceAPI`` operations.
+
+    :param unicode blockdevice_id: The unique identifier of the block device.
+    """
+    def __init__(self, blockdevice_id):
+        if not isinstance(blockdevice_id, unicode):
+            raise TypeError(
+                'Unexpected blockdevice_id type. '
+                'Expected unicode. '
+                'Got {!r}.'.format(blockdevice_id)
+            )
+        Exception.__init__(self, blockdevice_id)
+        self.blockdevice_id = blockdevice_id
+
+
+class UnknownVolume(VolumeException):
+    """
+    The block device could not be found.
+    """
+
+
+class AlreadyAttachedVolume(VolumeException):
+    """
+    A failed attempt to attach a block device that is already attached.
+    """
+
+
+class UnattachedVolume(VolumeException):
+    """
+    An attempt was made to operate on an unattached volume but the operation
+    requires the volume to be attached.
+    """
+
+
+class IBlockDeviceAPI(Interface):
+    """
+    Common operations provided by all block device backends.
+    """
+    def create_volume(size):
+        """
+        Create a new block device.
+
+        :param int size: The size of the new block device in bytes.
+        :returns: A ``BlockDeviceVolume``.
+        """
+
+    def attach_volume(blockdevice_id, host):
+        """
+        Attach ``blockdevice_id`` to ``host``.
+
+        :param unicode blockdevice_id: The unique identifier for the block
+            device being attached.
+        :raises UnknownVolume: If the supplied ``blockdevice_id`` does not
+            exist.
+        :raises AlreadyAttachedVolume: If the supplied ``blockdevice_id`` is
+            already attached.
+        :returns: A ``BlockDeviceVolume`` with a ``host`` attribute set to
+            ``host``.
+        """
+
+    def list_volumes():
+        """
+        List all the block devices available via the back end API.
+
+        :returns: A ``list`` of ``BlockDeviceVolume``s.
+        """
+
+    def get_device_path(blockdevice_id):
+        """
+        Calculate the path at which ``blockdevice_id`` will be exposed on the
+        host when attached.
```

Comment:
I think I've changed my mind again. What confused me was that we were creating the loopback device here in our `LoopbackBlockDeviceAPI`, which seems unrealistic.

With libcloud interface to OpenStack blocks, you have to supply the device path when you attach the device.
But that can be chosen automatically by the OpenStack implementation of BlockDeviceAPI.

Then `OpenStackBlockDeviceAPI.get_device_path` will be able to get that path from the libcloud API.